core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -538,6 +539,93 @@ class ReplicaManager(val config: KafkaConfig,
     }
   }
 
+  /*
+   * For each pair of partition and log directory specified in the map, record the pair in the memory so that the partition
+   * will be created in the specified log directory when broker receives LeaderAndIsrRequest for the partition later.
+   *
+   * This API is currently only useful if the replica has not been created yet. We will be able to move replicas
+   * that are already created to the user-specified log directory after KIP-113 is fully implemented
+   *
+   */
+  def alterReplicaDir(partitionDirs: Map[TopicPartition, String]): Map[TopicPartition, Errors] = {
+    partitionDirs.map { case (topicPartition, destinationDir) =>
+      try {
+        if (!logManager.isLogDirOnline(destinationDir))
+          throw new KafkaStorageException(s"Log directory $destinationDir is offline")
+
+        // If the log for this partition has not been created yet:
+        // 1) Respond with ReplicaNotAvailableException for this partition in the AlterReplicaDirResponse
+        // 2) Record the destination log directory in the memory so that the partition will be created in this log directory
+        //    when broker receives LeaderAndIsrRequest for this partition later.
+        getReplica(topicPartition) match {
+          case Some(_) => // The support for moving replica between log directories on the same broker is not available yet.
+          case None =>
+            logManager.updatePreferredLogDir(topicPartition, destinationDir)
+            throw new ReplicaNotAvailableException(s"Replica $localBrokerId is not available for partition $topicPartition")
```

Comment:
Hmm, intuitively, if a replica doesn't exist, setting a log dir for it should succeed, instead of getting an error. Perhaps we should return success if the replica doesn't exist or is in the right dir, and throw an unsupported exception for now otherwise.

Reply:
I think it depends on the semantics of AlterReplicaDirRequest. In my understanding "alter" means "change the proper of something that already exists". Thus broker should not create replica for this partition if this replica does not already exist. And it is reasonable to throw exception to user because user requested to alter log directory of a replica that doesn't exist.

Does this make sense?

Reply:
Hmm, I think AlterReplicaDirRequest just means the intention to alter the dir. It doesn't mean the change has to be completed. As long as the intention is remembered by the broker, it seems it's reasonable to return success. The weird thing right now is that in ReassignPartitionsCommand, the happy path is actually based on an exception in the response of AlterReplicaDirRequest. Normally, the happy path should be when the response has no error.

Reply:
I agree it doesn't have to be completed as long as it is remembered. However, currently this is only remembered in the memory which may be lost if broker restarts after it receives AlterReplicaDirRequest but before it receives LeaderAndIsrRequest to create the replica. I think we probably don't want to return success to user and create replica is a different log directory later (if restart happens in the above case).

@becketqin has similar comment regarding the weirdness that we expect response to throw exception. But this weirdness exists only because this patch represents the first part of KIP-113. After we fully implement the KIP-113, we won't expect to see exception in the happy path. The logic in the reassignment will look like this:

- Call `adminClient.alterReplicaDirRequest(replicaAssignment)` without waiting for response
- Create the reassignment znode so that controller can start replica reassignment across brokers.
- Call `adminClient.alterReplicaDirRequest(replicaAssignment)` and verify that there is no error in the response. The implementation of `adminClient.alterReplicaDirRequest` will treat `ReplicaNotAvailableException` as a retriable error and retry up to the user-specified timeout.

I think it is OK for this weird to exist for a short period of time before KIP-113 is fully implemented. User won't be affected by this weirdness as well. Does this make sense?

Reply:
Hmm, I thought for verification, we want to use describeLogDirs not alterReplicaDirRequest? Also, for the first adminClient.alterReplicaDirRequest(), I am not sure that we want to completely ignore the response. For example, if the request fails with authentication error, we probably want to error out, right?

Overall, are you saying that AlterReplicaDirRequest only returns no error if the log dir is in the target log dir? That means most of the time, the request will return error. This seems unintuitive since error should be the exception, not the norm.

Reply:
Yes, we only use `describeLogDirs` for verification. In the current patch, `ReassignPartitionsCommand` only checks for `ReplicaNotAvailableException` using `alterReplicaDirRequest` during execution (i.e. when --execute is specified).

You are right. I simplified the logic of the first `adminClient.alterReplciaDir()` in my previous response. More specifically, after KIP-113 is fully implemented, the `ReassignPartitionsCommand` should use `adminClient.alterReplciaDir()` to send `AlterReplicaDirRequest` without retry. And it should verify that either there is no error or the error is `ReplicaNotAvailableException`. The second `adminClient.alterReplciaDir()` should retry `AlterReplicaDirRequest` upon `ReplicaNotAvailableException` until timeout.

`AlterReplicaDirRequest` can also return no error even if the log directory is not the target log directory, as long as replica already exists on the broker. This is because if the replica already exists on the broker, the broker will create a directory for the temporary replica in the destination log directory. Because information is persisted on the disk rather than in memory, broker will continue to move replica to the destination log directory after restart.

Reply:
Ok, thanks for the explanation. This is fine then. Could we document the error code/exception in AlterReplicaDirResult?

Reply:
@junrao Sure. Previously the error is only documented in `AlterReplicaDirResponse`. Just now I added the error code documentation in `AlterReplicaDirResult` as follows:

```
* LOG_DIR_NOT_FOUND (57)
* KAFKA_STORAGE_ERROR (56)
* REPLICA_NOT_AVAILABLE (9)
* UNKNOWN (-1)
```